security/advancedtls/advancedtls.go

```diff
@@ -54,17 +54,34 @@ type VerificationFuncParams struct {
 	Leaf *x509.Certificate
 }
 
-// VerificationResults contains the information about results of
-// CustomVerificationFunc.
-// VerificationResults is an empty struct for now. It may be extended in the
+// DEPRECATED: Renamed to `HandshakeVerificationInfo`
+// VerificationFuncParams contains parameters available to users when
+// implementing CustomVerificationFunc.
+// The fields in this struct are read-only.
+type VerificationFuncParams = HandshakeVerificationInfo
+
+// PostHandshakeVerificationResults contains the information about results of
+// PostHandshakeVerificationFunc.
+// PostHandshakeVerificationResults is an empty struct for now. It may be extended in the
 // future to include more information.
-type VerificationResults struct{}
+type PostHandshakeVerificationResults struct{}
+
+// DEPRECATED: Renamed to `PostHandshakeVerificationResults`
+type VerificationResults = PostHandshakeVerificationResults
+
+// PostHandshakeVerificationFunc is the function defined by users to perform
+// custom verification checks after chain building and regular handshake
+// verification has been completed.
+// PostHandshakeVerificationFunc returns nil
+// if the authorization fails; otherwise returns an empty struct.
+type PostHandshakeVerificationFunc func(params *HandshakeVerificationInfo) (*PostHandshakeVerificationResults, error)
```

Comment:
Suggestion: `PeerVerificationFunc` (and all associated types)?  It's a shorter name and (maybe?) is named after the intent of the thing vs. when it happens.

Reply:
We want to make sure that we are differentiating between verification happening after the normal chain building and verification that happens by default (which this is), and overriding the base chain building and verification itself.

Reply:
I'm not sure I follow all this.

What is "post handshake verification" exactly?  And you're saying this can be used to _replace_ the default behavior?  What is the default behavior?  Do you have any examples of what users would want to do with this?

Reply:
What's the API for (2)?  `VerifyPeer`?

So is it the case that if you specify this callback then `VerifyPeer` isn't called?

Reply:
Renamed `VerifyPeer` in the settings struct, PTAL